### PR TITLE
feat(calling): Group calling

### LINKED ISSUE #393: IsTalking should be activated on the profile picture when a users incoming stream has audio attached.
When making a call if the user is talking there should be an indicator behind them using the isTalking prop on the profile picture. this same effect should be applied to the outside of videos when the person is talking.

### PR DESCRIPTION
**Pull request by Flemmli97**
### What this PR does 📖

- Implements group calling:
  - When starting a call a room will be created and invitations will be send out to all users in chat.
  - If a user leaves the call the call will be ended if only one person is left in it
- Also fixes camera/mic not properly closed for calling, audio settings and video test in dev settings

### Which issue(s) this PR fixes 🔨

- Resolve #393

### Notes
- UI rework needs to be still done
- Video is not toggable from the other users end (the caller can toggle). But idk if this is due to my setup where i test it on the same computer or not. Also see #466
- It takes a few second when accepting a call to join the room fully

### CODE REVIEW

**Comment by Flemmli97** [comment]
> Something I noticed while testing (I can make this into a new ticket if we just wanna get this merged in)
> 
> If User A calls a group with User B and User C and only User B answers the call, it will still look like UserC is in the call with them even though he is not
> 
> Like I said lemme know if this would be better handled in a new ticket and I will be happy to create one!
> 
>  Screen_Recording_2024-09-17_at_4.52.49_PM.mov

changed that now

**Comment by Flemmli97** [comment]
> Looks better but just found one thing that happens now after the fix
> 
> When UserA/B are in the call and UserC deny's it, once UserC hits deny the ringtone will re-appear and the user will need to deny it again
> 
>  Screen.Recording.2024-09-18.at.11.50.10.AM.mov

are there any other problems? does the main thing all work?

**Review by lgmarchi — APPROVED** [review]

**Comment by Flemmli97** [comment]
> > > Looks better but just found one thing that happens now after the fix
> > > When UserA/B are in the call and UserC deny's it, once UserC hits deny the ringtone will re-appear and the user will need to deny it again
> > > Screen.Recording.2024-09-18.at.11.50.10.AM.mov
> > 
> > 
> > are there any other problems? does the main thing all work?
> 
> Seems for the most part general calling works. If we want once the conflicts are fixed I will approve and create tickets for any issues

i will fix the issue above in this pr still

File: src/lib/media/Voice.ts [hunk 1]
```diff
@@ -45,352 +49,675 @@ type VoiceMessage = {
...
+                await new Promise(resolve => setTimeout(resolve, 2000))
+                if (!connected) {
+                    if (attempts === maxRetries) {
+                        log.error(`Connection to ${peer.remotePeerId} failed.`)
+                        return false
+                    }
+                    log.debug(`Not possible to connect to ${peer.remotePeerId}, trying again.`)
+                    continue
+                }
+                // Call after successfuly connect. But dont wait for it
```
Comment (stavares843) [inline src/lib/media/Voice.ts:226]:
```suggestion
                // Call after successfully connect. But dont wait for it
```

File: src/lib/components/calling/CallScreen.svelte [hunk 2]
```diff
@@ -140,32 +133,47 @@
...
         }
 
-        if (VoiceRTCInstance.remoteVideoElement) {
-            remoteVideoElement.srcObject = VoiceRTCInstance.activeCall?.remoteStream!
-            remoteVideoElement.play()
-        }
+        // if (VoiceRTCInstance.remoteVideoElement) {
+        //     remoteVideoElement.srcObject = VoiceRTCInstance.remotePeersHolder.remotePeers[0]?.activeCall?.remoteStream!
+        //     remoteVideoElement.play()
+        // }
```
Comment (stavares843) [inline src/lib/components/calling/CallScreen.svelte:174]: do we need this?

File: src/lib/media/Voice.ts [hunk 3]
```diff
@@ -3,22 +3,26 @@ import { SettingsStore } from "$lib/state"
...
 import { _ } from "svelte-i18n"
+import { get } from "svelte/store"
 
 export enum VoiceRTCMessageType {
     Calling = "CALLING_USER",
-    EndingCall = "ENDING_CALL",
-    IncomingCall = "INCOMING_CALL",
-    AcceptedCall = "ACCEPTED_CALL",
+    LeavingCall = "LEAVING_CALL",
+    JoinedCall = "JOINED_CALL",
```
Comment (stavares843) [inline src/lib/media/Voice.ts:13]: should this be JOINING? to be joining/leaving

File: src/lib/media/Voice.ts [hunk 4]
```diff
@@ -45,352 +49,675 @@ type VoiceMessage = {
...
+        this.streamHandler = [checker, analyser]
+    }
 
-        this.callOptions.call.onlyAudioCall = conn.metadata.onlyAudioCall
-        this.callOptions.video.enabled = !conn.metadata.onlyAudioCall && this.callOptions.video.enabled
+    async handleRemoteStream(stream: MediaStream, remove: ((user: string) => void) | undefined) {
+        this.handleStreamMeta(stream)
+        this.stream = stream
+        /// Here will receive data from user that made the call
+        // TODO: decide which video to play?
```
Comment (stavares843) [inline src/lib/media/Voice.ts:405]: maybe we could by default show the video of who started the call

File: src/lib/media/Voice.ts [hunk 5]
```diff
@@ -443,56 +767,54 @@ export class VoiceRTC {
...
-        const processedStream = new MediaStream()
-        processedStream.addTrack(destination.stream.getAudioTracks()[0])
-        processedStream.addTrack(this.activeCall?.localStream!.getVideoTracks()[0]!)
-        this.activeCall?.localStream!.getVideoTracks().forEach(track => processedStream.addTrack(track))
-    }
+    //     const processedStream = new MediaStream()
+    //     processedStream.addTrack(destination.stream.getAudioTracks()[0])
+    //     processedStream.addTrack(this.activeCall?.localStream!.getVideoTracks()[0]!)
+    //     this.activeCall?.localStream!.getVideoTracks().forEach(track => processedStream.addTrack(track))
+    // }
```
Comment (stavares843) [inline src/lib/media/Voice.ts:796]: do we need this?

File: src/lib/media/VoiceO.ts [hunk 6]
```diff
@@ -0,0 +1,625 @@
+import { CallDirection } from "$lib/enums"
```
Comment (stavares843) [inline src/lib/media/VoiceO.ts:1]: do we need this file? src/lib/media/VoiceO.ts or was this for testing?


if so, can we update the naming?

File: src/routes/+layout.svelte [hunk 7]
```diff
@@ -195,7 +195,7 @@
         <MouseListener on:clicked={() => {}} />
         <Toasts />
         <IncomingCall />
-        <VideoPreview />
+        <!-- <VideoPreview /> This needs a rework as atm it only supports one to one calling. Maybe add ability to select source? -->
```
Comment (InfamousVague) [inline src/routes/+layout.svelte:198]: Ideally this should show the video of whoever is talking at the time, I know this is more involved so no worries about this in this PR but let's get a ticket opened up so we don't forget about it

**Review by InfamousVague — APPROVED** [review]

**Review by stavares843 — COMMENTED** [review]
@phillsatellite and i were testing the PR

feedback

Phil's Call Issue (Unanswered Modal):
Scenario: Phil started a call, and one of the users saw the call modal appear but was unable to answer it.

https://github.com/user-attachments/assets/572ee9ed-5abb-4565-bcac-73f28287ca65


User Count Discrepancy:
Scenario: During a call, only one user had joined, but the system indicated that three users were present.

image


Unknown User on Call Modal:
Scenario: When Phil called Sara, the call modal displayed "Unknown User" instead of Sara's name.

image


No Call Modal Appearing:
Scenario: Phil called Sara, but no call modal appeared on her side.

Ringtone Without Modal:
Scenario: Phil called Sara, and while she heard the ringtone, the call modal did not appear.

Call Not Starting or Missed Call Notifications:
Scenario: Sara attempted to initiate an audio or video call, but it didn’t start. Phil then called Sara, but no call appeared for her, nor did missed call notifications show up in the chat. After refreshing the app, the issue was resolved.

Phil's Call Issue (No Calls Initiated):
Scenario: Phil attempted to call Sara, but the call was not initiated.

File: src/lib/media/Voice.ts (hunk 5 above)
Comment (Flemmli97) [reply src/lib/media/Voice.ts:807]: its from dev. but unused atm

**Comment by Flemmli97** [comment]
can you post the logs for the first problem?
and for the rest have you tried the up to date code? (i also pushed some otherstuff just now)

**Comment by stavares843** [comment]
we don't have logs for the first issue - we can try to reproduce and get logs

but yes we had the branch updated on our side, new account, clean builds and everything

File: src/lib/media/Voice.ts [hunk 8]
```diff
@@ -43,354 +47,685 @@ type VoiceRTCUser = {
...
+            log.error("Could not connect to anyone")
+        }
     }
 
-    toggleDeafen(state: boolean) {
-        // TODO: This isn't perfect because if you mute yourself, and then deafen yourself, un-deafaning will also unmute you which could be unexpected
-        this.callOptions.audio.enabled = state
+    private async connectWithRetryFor(localPeer: Peer, channel: string, callOptions: VoiceRTCOptions, peer: RemotePeer) {
+        if (this.connectedPeers.has(peer.remotePeerId)) return true
+        const maxRetries = 5 // TODO: Failing to connect after 2 attemps (or just throwing errors twitce for the local peer?) breaks ALL connections...
```
Comment (stavares843) [inline src/lib/media/Voice.ts:189]:
```suggestion
        const maxRetries = 5 // TODO: Failing to connect after 2 attempts (or just throwing errors twice for the local peer?) breaks ALL connections...
```
```suggestion
        const maxRetries = 5 // TODO: Failing to connect after 2 attemps (or just throwing errors twitce for the local peer?) breaks ALL connections...
```

**Comment by stavares843** [comment]
we will try again with your latest commit to see if they still happen

**Comment by stavares843** [comment]
so @phillsatellite and i retested with latest commit

Phil called Sara. When Sara answered, the call didn't appear on her side, but Phil saw that Sara was on the call. 

Later, Sara called Phil, and though Phil successfully joined the call, Sara couldn't see Phil on the line, while Phil saw her as being connected.



Phil side

image


Sara side

image

**Comment by luisecm** [comment]
@Flemmli97 @stavares843 Looks like the build android job is failing on this branch - https://github.com/Satellite-im/UplinkWeb/actions/runs/11036050320/job/30653555494?pr=565

**Review by InfamousVague — APPROVED** [review]

**Comment by phillsatellite** [comment]
Noticed some things while testing the latest 

When the call came in for User B & C it appeared as though it was coming from an unknown user 

All the users joined the call but on User A & C side it says only 2 users are in the call. For user B it showed only themselves in the call 

When User A hangs up it also hung up the call for User B. User C still remains in the call 


https://github.com/user-attachments/assets/61b667db-4445-4497-8a6c-4c9e91c0cc75

Screenshot 2024-09-26 at 11 40 02 AM

**Comment by Flemmli97** [comment]
what happens if you wait a bit longer? sometimes i had the problem of needing to wait quite a few seconds. even for 1 to 1 call

**Comment by Flemmli97** [comment]
> Noticed some things while testing the latest
> 
> When the call came in for User B & C it appeared as though it was coming from an unknown user
> 
> All the users joined the call but on User A & C side it says only 2 users are in the call. For user B it showed only themselves in the call
> 
> When User A hangs up it also hung up the call for User B. User C still remains in the call
> 
>  Screen.Recording.2024-09-26.at.11.37.33.AM.mov 
> 
> Screenshot 2024-09-26 at 11 40 02 AM

i fixed various stuff. but as for users not showing up:
it has to do with multipass failing to find the identity at the start.
i can implement a retry system there but since that is not directly involved with calling would it be better to do it in a new PR?

**Comment by stavares843** [comment]
we will test again

**Comment by Flemmli97** [comment]
check with #653 to see if its still a problem with users not showing up during call

**Comment by Flemmli97** [comment]
also one more issue that might be resolved is #463. at least using this branch i dont have that problem

**Comment by phillsatellite** [comment]
> also one more issue that might be resolved is #463. at least using this branch i dont have that problem

@Flemmli97 both of these issues do not happen in #653 💥

**Review by stavares843 — CHANGES_REQUESTED** [review]
phil called me and i called phil both group and 1:1 and always say 1 user on call


image

**Comment by Flemmli97** [comment]
hmm can you post the whole console log?

**Comment by phillsatellite** [comment]
Screenshot 2024-10-02 at 1 00 40 PM

**Comment by Flemmli97** [comment]
can you try add this line in Voice.ts `console.log("own did: ", get(Store.state.user).key)` in line 452? before the connect call. and then see what it prints. this is on the side of the caller.
cause from the pic you send it seems its sending an undefined did over
maybe an account is corrupt?

**Comment by Flemmli97** [comment]
> can you try add this line in Voice.ts `console.log("own did: ", get(Store.state.user).key)` in line 452? before the connect call. and then see what it prints. this is on the side of the caller. cause from the pic you send it seems its sending an undefined did over
maybe an account is corrupt?

**Comment by phillsatellite** [comment]
> can you try add this line in Voice.ts `console.log("own did: ", get(Store.state.user).key)` in line 452? before the connect call. and then see what it prints. this is on the side of the caller. cause from the pic you send it seems its sending an undefined did over maybe an account is corrupt?

Will give it a shot. And shouldnt be a corrupt account these were all brand new

**Comment by phillsatellite** [comment]
@Flemmli97 Sara and I retested and it is now showing the correct amount of users in the call. There just seems to be another issue that is appearing now. 

When you are in a group call there is maybe 5 seconds in the beginning of the call when the sound works and after that 5 seconds there is no sound even if you try to mute/unmute or deafen/undeafen users cannot hear each other at all. On dev branch audio on calls are fine 

Screenshot 2024-10-02 at 5 42 15 PM

**Review by stavares843 — APPROVED** [review]

**Pull request merged**